File: BinarySubtraction.py
def DecimalToBinary(n, num):
    bin = ""
    while(num>0):
        rem = num % 2
        num = num // 2
        bin +=  "".join(str(rem))
    return bin+"0"*(n-len(bin))

def BinaryToDecimal(bin):
    Decimal = 0
    for i in range(len(bin)):
        Decimal += int(bin[i])*(2**i)
    return Decimal

def OnesComplement(bin):
    cBin =""
    for ele in bin:
        if ele == "0":
            cBin += "".join("1")
        else:
            cBin += "".join("0")
    return cBin
    
def BinaryAddition(bin1, bin2):
    result, carry = "", 0
    for i in range(len(bin1)):
        sum = int(bin1[i]) ^ int(bin2[i]) ^ carry
        carry = (int(bin1[i]) & int(bin2[i])) | (int(bin2[i]) & carry) | (carry & int(bin1[i]))
        result += "".join(str(sum))
    return result

if __name__ == '__main__':
    num = list(map(int, input("Enter two decimal numbers: ").split(" ")))
    
    Binary1 = DecimalToBinary(8, num[0])
    print(f"Binary of {num[0]} is: ", Binary1[::-1])
    
    Binary2 = DecimalToBinary(8, num[1])
    print(f"Binary of {num[1]} is: ", Binary2[::-1])
    
    OneCmp = OnesComplement(Binary2)
    print("1's complement: ", OneCmp[::-1])
    
    TwoCmp = BinaryAddition(OneCmp, "10000000")
    print("2's complement: ", TwoCmp[::-1])
    
    Difference = BinaryAddition(Binary1, TwoCmp)
    print(f"Difference of {num[0]} and {num[1]} is: ", BinaryToDecimal(Difference))
    
    
    
    

File: BinarySubtraction.cpp
#include <iostream>
#include <math.h>
using namespace std;

void BinaryAddition(int, int[], int[], int[]);
void DecimalToBinary(int, int []); // decimal number for binary conversion and array to store it.
void BinaryToDecimal(int, int []);
void Display(int n, int []);
void OnesComplement(int, int []);

int main(){
    int deci1=0, deci2=0;
    int arr1[8] = {0,0,0,0,0,0,0,0}; // initilization is done to ensure array gets garbage free value.
    int arr2[8] = {0,0,0,0,0,0,0,0};
    int arr3[8] = {1,0,0,0,0,0,0,0};
    int result[8] = {0,0,0,0,0,0,0,0}; // stores 2's complement
    int difference[8] = {0,0,0,0,0,0,0,0}; // stores the differece 
    cout << "enter two decimal numbers(0-255): ";
    cin >> deci1 >> deci2;

    // convert first decimal number to binary and display
    DecimalToBinary(deci1, arr1);
    cout << "Binary equivalent of " << deci1 << " is: ";
    Display(8, arr1); // 8 bit 

    // Convert second decimal number to binary and display
    DecimalToBinary(deci2, arr2);
    cout << "\nBinary Equivalent of " << deci2 << " is: ";
    Display(8, arr2); // 8-bit

    // find 1's complement of subtrahend(smaller number)
    OnesComplement(8, arr2);
    cout << "\nThe 1's complement is: ";
    Display(8, arr2);

    // Add 1 to 1's complement of subtrahend to find 2's complement
    BinaryAddition(8, arr2, arr3, result); // 8-bit, two array for add and other array to store result
    cout << "\n2's Complement equivalent is: ";
    Display(8, result);
    
    //Add Minuend and 2's complement of Subtrahend
    BinaryAddition(8, arr1, result, difference);
    cout<< "\nThe differece equivalent in binary: ";
    Display(8, difference);

    //convert difference of two number to decimal    
    BinaryToDecimal(8, difference);

    return 0;
}

// Funtion to convert decimal to binary.
void DecimalToBinary(int num, int arr[]){
    int i = 0, rem = 0;
    do{
        rem = num % 2;
        arr[i] = rem;
        num = num / 2;
        i++;
    }while(num > 0);
}

// Function to display the array
void Display(int n, int arr[]){
    int i;
    for(i=n-1;i>=0;i--){
        cout << arr[i];
    }
}

//Function to convert binary to decimal
void BinaryToDecimal(int n, int arr[]){
    int i, result=0;
    for(i=0;i<n;i++){
        result += arr[i]*pow(2,i);
    }
    cout << "\nDecimal Equivalent of difference is: " << result << endl;
}

// Function to add two binary numbers
void BinaryAddition(int n, int arr1[], int arr2[], int result[]){
    int carry=0, i=0;
    for(i=0;i<n;i++){
        result[i] = (arr1[i] ^ arr2[i] ^ carry);
        carry = (arr1[i] & arr2[i]) | (arr2[i] & carry) | (carry & arr1[i]);
    }
}

// Function to find 1's complement
void OnesComplement(int n, int arr[]){
    for(int i=0;i<n;i++){
        if(arr[i]==0) arr[i] = 1;
        else arr[i] = 0;
    }
}
